src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1156,6 +1156,46 @@ void MergeTreeData::clearOldTemporaryDirectories(ssize_t custom_directories_life
     }
 }
 
+MergeTreeData::DataPartsVector MergeTreeData::grabOldModifiedParts()
+{
+    DataPartsVector parts;
+
+    if (grab_old_modified_parts_called)
```

Comment:
This is not the way to use atomics. You have a race condition in here (line 1163 vs line 1167). You need to employ the class's atomic operations, like `grab_old_modified_parts_called.exchange(true)`.